### ISSUE TITLE
distinct() issue

### ISSUE THREAD

**Issue by RJFares** [realm/realm-java#2947]
#### Goal

Get a distinct list of a RealmResult
#### Expected Results

RealmResult.distinct() returning a distinct list
#### Actual Results

RealmResult.distinct() sorting in-place
#### Steps & Code to Reproduce

mList = mRealm.where(Item.class).findAll();
mDistinctList = mList.distinct(Item.CATEGORY)

The 2nd line of code above will change mList also (is that intentional or a bug?)
#### Version of Realm and tooling

Realm version(s): 1.1.0
Android Studio version: 2.1.1

*\* EDIT *\* the above issue happens in RealmResults only and not RealmLists

**Comment by stk1m1** [comment]
@RJFares 
As of now, it works as designed. We have addressed the issue #2260, and been looking into it.

When you're in need of passing multiple arguments for distinctive elements, you can use [`distinct(...)`](https://realm.io/docs/java/1.0.0/api/io/realm/RealmResults.html#distinct-java.lang.String-java.lang.String...-) such as below.

```java
mDistinctList = mRealm.where(Item.class)
                      .distinct(Item.CATEGORY_1, Item.CATEGORY_2, ...)
                      .findAll();
```

**Comment by Zhuinden** [comment]
@stk1m1 if `sort()` returns a new copy of `RealmResults<T>` then `distinct()` should also return a new copy rather than modify the results in place

**Comment by stk1m1** [comment]
Thanks for the input, @Zhuinden!

The way `distinct()` works has been actually discussed, and a patch is on its way. It's just that others have higher priority that we haven't had an opportunity to polish and release.

**Comment by vishal901** [comment]
RealmResults<Abc> results = realm.where(Abc.class).findAll();
  RealmResults<Abc> Abc= results.distinct("Name");

work done..............

**Comment by RJFares** [comment]
that is not the issue here.
If in a fragment that hosts a recyclerview with an adapter linked to RealmResults mList, I decide I want a separate .distinct list from mList, 

I cannot do:
RealmResults mDistinctList = mList.distinct("name");

because this will change mList also. The solution is to duplicate mList to mList2 then calling distinct and grabbing them with mDistinctList (otherwise the data in recyclerview will change).

The only drawback is that I'm creating an extra useless List. Realm is fast and memory will not be a big issue,but I'm wondering if distinct's behavior is intended like this or not.

**Comment by Zhuinden** [comment]
I think you can do `.where().findAll().distinct()` but I still think this is a bug or at least inconsistent behavior

**Comment by RJFares** [comment]
This assumes I need a findAll() or that search parameters remain constant. 
Unfortunately the contents of mList could have different search parameters, like sorting ascending or descending, or even changing a search field for example.

**Comment by Zhuinden** [comment]
I mean calling this series of operations on the results themselves... But I am not sure about it.

**Comment by Zhuinden** [comment]
This was closed in 3.0.0

**Comment by cmelchior** [comment]
Your right, thanks 👍
